Introduce torch.onnx.dynamo_export API

abock opened the pull request:
This is the first phase of the new ONNX exporter API for exporting from TorchDynamo and FX, and represents the beginning of a new era for exporting ONNX from PyTorch.

The API here is a starting point upon which we will layer more capability and expressiveness in subsequent phases. This first phase introduces the following into `torch.onnx`:

```python
dynamo_export(
    model: torch.nn.Module,
    /,
    *model_args,
    export_options: Optional[ExportOptions] = None,
    **model_kwargs,
) -> ExportOutput:
    ...

class ExportOptions:
    opset_version: Optional[int] = None
    dynamic_shapes: Optional[bool] = None
    logger: Optional[logging.Logger] = None

class ExportOutputSerializer(Protocol):
    def serialize(
        self,
        export_output: ExportOutput,
        destination: io.BufferedIOBase,
    ) -> None:
        ...

class ExportOutput:
    model_proto: onnx.ModelProto

    def save(
        self,
        destination: Union[str, io.BufferedIOBase],
        *,
        serializer: Optional[ExportOutputSerializer] = None,
    ) -> None:
        ...
```

In addition to the API in the first commit on this PR, we have a few experiments for exporting Dynamo and FX to ONNX that this PR rationalizes through the new Exporter API and adjusts tests to use the new API.

- A base `FXGraphModuleExporter` exporter from which all derive:
  - `DynamoExportExporter`: uses dynamo.export to acquire FX graph
  - `DynamoOptimizeExporter`: uses dynamo.optimize to acquire FX graph
  - `FXSymbolicTraceExporter`: uses FX symbolic tracing

The `dynamo_export` API currently uses `DynamoOptimizeExporter`.

### Next Steps (subsequent PRs):

* Combine `DynamoExportExporter` and `DynamoOptimizeExporter` into a single `DynamoExporter`.
* Make it easy to test `FXSymbolicTraceExporter` through the same API; eventually `FXSymbolicTraceExporter` goes away entirely when the Dynamo approach works for large models. We want to keep `FXSymbolicTraceExporter` around for now for experimenting and internal use.
* Parameterize (on `ExportOptions`) and consolidate Dynamo exporter tests.
  - This PR intentionally leaves the existing tests unchanged as much as possible except for the necessary plumbing.
* Subsequent API phases:
  - Diagnostics
  - Registry, dispatcher, and Custom Ops
  - Passes
  - Dynamic shapes

Fixes #94774

Commit:
Introduce torch.onnx.dynamo_export API

This is the first phase of the new ONNX exporter API for exporting from
TorchDynamo and FX, and represents the beginning of a new era for exporting
ONNX from PyTorch.

The API here is a starting point upon which we will layer more capability
and expressiveness in subsequent phases. This first phase introduces the
following into `torch.onnx`:

dynamo_export(
    model: torch.nn.Module,
    /,
    *model_args,
    export_options: Optional[ExportOptions] = None,
    **model_kwargs,
) -> ExportOutput

class ExportOptions:
    opset_version: Optional[int] = None
    dynamic_shapes: Optional[bool] = None
    logger: Optional[logging.Logger] = None

class ExportOutputSerializer(Protocol):
    def serialize(
        self,
        export_output: ExportOutput,
        destination: io.BufferedIOBase
    ) -> None:
        ...

class ExportOutput:
    export_options: ExportOptions
    model_proto: onnx.ModelProto

    def save(
        self,
        destination: Union[str, io.BufferedIOBase],
        *,
        serializer: Optional[ExportOutputSerializer] = None
    ) -> None:
        ...

Co-authored-by: Bowen Bao <bowbao@microsoft.com>
Co-authored-by: Aaron Bockover <abock@microsoft.com>

justinchuby commented on torch/onnx/_internal/exporter.py:
```diff
@@ -0,0 +1,248 @@
+
+    resolved_export_options = _resolve_export_options(export_options)
+
+    def assert_dependencies():
```
Curious what are the messages we display to users when dependencies are missing?

abock replied:
This depends on Phase 2: Diagnostics. @BowenBao will own that. Messages will be granular and actionable.

justinchuby commented on torch/onnx/_internal/exporter.py:
```diff
@@ -0,0 +1,248 @@
+    def serialize(
+        self, export_output: ExportOutput, destination: io.BufferedIOBase
+    ) -> None:
+        pass
```
```suggestion
        ...
```

optional: `...` tends to be conventional for protocols I think?

abock commented on torch/onnx/_internal/exporter_impl.py:
```diff
@@ -0,0 +1,13 @@
+    def run(self) -> ExportOutput:
+        from pprint import pprint
+
+        pprint(self)
```
No, `pprint` is nicer for dumping objects. More importantly, this is also code that is intended to be replaced entirely (see "Next Steps" in the PR description).

justinchuby commented on torch/onnx/errors.py:
```diff
@@ -46,6 +45,9 @@ class UnsupportedOperatorError(OnnxExporterError):
 
     def __init__(self, name: str, version: int, supported_version: Optional[int]):
+        # Delayed import to avoid circular dependency.
+        from torch.onnx._internal import diagnostics
```
@BowenBao I wonder what caused this?

abock replied:
Beartype.

BowenBao replied:
beartype -> errors -> diagnostics -> beartype

BowenBao replied:
Technically we can solve this in another way by moving down the new imports
```python
from ._internal.exporter import (
    dynamo_export,
    ExportOptions,
    ExportOutput,
    ExportOutputSerializer,
)
```
to after
```python
from . import (  # usort:skip. Keep the order instead of sorting lexicographically
    _deprecation,
    errors,
    symbolic_caffe2,
    symbolic_helper,
    symbolic_opset7,
    symbolic_opset8,
    symbolic_opset9,
    symbolic_opset10,
    symbolic_opset11,
    symbolic_opset12,
    symbolic_opset13,
    symbolic_opset14,
    symbolic_opset15,
    symbolic_opset16,
    symbolic_opset17,
    symbolic_opset18,
    utils,
)
```
May need a `# usort:skip` tag if lintrunner is not happy with it. I don't have strong preference either way.

thiagocrepaldi replied:
Probably the best fix is making beartype wrapper more robust so that users don't hit this from other places too

titaiwangms reviewed (approved):
Good work

BowenBao commented on torch/onnx/_internal/exporter.py:
```diff
@@ -0,0 +1,248 @@
+should reference ``ExportOptions.opset_version``."""
+
+
+@_beartype.beartype
```
this needs to move to `__init__`, now that `ExportOptions` is no longer `dataclass`.

justinchuby replied:
Looks like we can still decorate a class: https://beartype.readthedocs.io/en/latest/api_decor/#class-mode

thiagocrepaldi commented on torch/onnx/_internal/exporter.py:
```diff
@@ -0,0 +1,248 @@
+        # - https://github.com/microsoft/onnx-script/issues/572
+        import onnxscript  # type: ignore[import]
+
+        assert onnx.defs.onnx_opset_version() >= opset_version
```
```suggestion
        assert onnx.defs.onnx_opset_version() >= opset_version, "The installed ONNX does not supported the specified opset version. Install a newer ONNX or specify an older opset version!"
```

thiagocrepaldi commented on torch/onnx/_internal/exporter.py:
```diff
@@ -0,0 +1,248 @@
+        assert isinstance(
+            onnxscript.onnx_opset.all_opsets[("", opset_version)],
+            onnxscript.values.Opset,
+        )
```
```suggestion
        ), "The installed ONNX Script does not support the specified opset version. Install a newer ONNX Script or specify an older opset version!"
```

BowenBao replied:
This can be part of the PR that plumb through the FX exporter. Adding implementation and hardening checks.

* Raise `ValueError`
* Raise with formatted error message with user specified version, and acceptable range.
* Update docstring with `Raises:`

thiagocrepaldi replied:
I am inclined to agree, but this PR is already adding some implementation. Just maikng sure that whatever is added, is added in the best possible way (giving the partial support for v1)

abock replied:
Updated.

thiagocrepaldi commented on torch/onnx/_internal/exporter.py:
```diff
@@ -0,0 +1,248 @@
+    """An in-memory representation of a PyTorch model that has been exported to ONNX."""
+
+    _export_options: Final[ExportOptions]
+    _model_proto: Final[onnx.ModelProto]
```
```suggestion
    _model_proto: Final[onnx.ModelProto]
    """Model proto used by ONNX spec to represent the model.
    
    NOTE: In future releases this will be replaced by an agnostic serialization graph IR to allow multiple serialization formats.
    """
    Maybe a note on why `ExportOutput` has a model_proto instead of it being produced by `ProtobufExportOutputSerializer`?
```

abock replied:
We have not actually committed to this yet. I do not want to call that out at all.

abock replied:
As for ExportOutput holding model_proto... the type's docstring itself says such. The serializer takes the in-memory representation and ... serializes it.

thiagocrepaldi replied:
> As for ExportOutput holding model_proto... the type's docstring itself says such. The serializer takes the in-memory representation and ... serializes it.

I got a little confused because the suggested serializer implementation is just a "write to disk something already proto-serialized by another component". It doesn't plumb the real serialization through torchscript/onnxscript. In this sense, the serializer is just a "disk-writer" of an already serialized model. That could be seen as an unnecessary layer and my comment was in the sense to clear the air on the what could be seen as an over design thing.

Have we not commited to a generic IR that is further serialized to protobuf (or whatever), we could completely eliminate the `ExportOutputSerializer` concept and do a simple `destination.write(self.model_proto.SerializeToString())` as `ExportOutput.save(self, *, destination: Union[str, io.BufferedIOBase])`. `ExportOutputSerializer` would only be introduced if we ever decide to support anything other than protobuf or decided that `ExportOutput` would not receive a `model_proto` as input.

Makes sense?

shubhambhokare1 commented on torch/onnx/_internal/exporter.py:
```diff
@@ -0,0 +1,284 @@
+
+    try:
+        # mypy is unhappy with 'import onnxscript' right now:
+        # - https://github.com/microsoft/onnx-script/issues/572
```
Should be fixed in https://github.com/microsoft/onnx-script/issues/572

shubhambhokare1 reviewed (approved):
LGTM 👍

abock committed:
Plumb various exporter strategies through new API

We have a few experiments for exporting Dynamo and FX to ONNX
that this PR rationalizes through the new Exporter API and
adjusts tests to use the new API.

- A base FXGraphModuleExporter exporter from which all derive:
- DynamoExportExporter: uses dynamo.export to acquire FX graph
- DynamoOptimizeExporter: uses dynamo.optimize to acquire FX graph
- FXSymbolicTraceExporter: uses FX symbolic tracing

The dynamo_export API currently uses DynamoOptimizeExporter.

titaiwangms commented on torch/onnx/_internal/exporter.py:
```diff
@@ -0,0 +1,309 @@
+    import onnx
+
+
+_DEFAULT_OPSET_VERSION: Final[int] = 18
```
I wonder if we could put this into torch.onnx._constants?

abock replied:
I'd like to keep it separate for now since `_constants` is for the torchscript path.

abock commented:
@pytorchbot merge

abock commented:
@pytorchbot merge
